qutebrowser/browser/history.py

```diff
@@ -288,14 +199,101 @@ def add_url(self, url, title="", *, redirect=False, atime=None):
         if atime is None:
             atime = time.time()
         entry = Entry(atime, url, title, redirect=redirect)
-        if self._initial_read_done:
-            self._add_entry(entry)
-            self._new_history.append(entry)
-            self.item_added.emit(entry)
-            if not entry.redirect:
-                self.add_completion_item.emit(entry)
+        self._add_entry(entry)
+
+    def _parse_entry(self, line):
+        """Parse a history line like '12345 http://example.com title'."""
+        data = line.split(maxsplit=2)
+        if len(data) == 2:
+            atime, url = data
+            title = ""
+        elif len(data) == 3:
+            atime, url, title = data
+        else:
+            raise ValueError("2 or 3 fields expected")
+
+        url = QUrl(url)
+        if not url.isValid():
+            raise ValueError("Invalid URL: {}".format(url.errorString()))
+
+        # https://github.com/qutebrowser/qutebrowser/issues/670
+        atime = atime.lstrip('\0')
+
+        if '-' in atime:
+            atime, flags = atime.split('-')
         else:
-            self._add_entry(entry, target=self._temp_history)
+            flags = ''
+
+        if not set(flags).issubset('r'):
+            raise ValueError("Invalid flags {!r}".format(flags))
+
+        redirect = 'r' in flags
+        row = (url, title, float(atime), redirect)
+        completion_row = None if redirect else (url, title, float(atime))
+
+        return (row, completion_row)
+
+    def import_txt(self):
+        """Import a history text file into sqlite if it exists.
+
+        In older versions of qutebrowser, history was stored in a text format.
+        This converts that file into the new sqlite format and removes it.
+        """
+        path = os.path.join(standarddir.data(), 'history')
+        if not os.path.isfile(path):
+            return
+
+        def action():
+            with debug.log_time(log.init, 'Import old history file to sqlite'):
+                self._read(path)
+                message.info('History import complete. Removing {}'
+                             .format(path))
+                os.remove(path)
```

Comment:
It only deletes if it succeeds, and then you have a db which you can restore with `debug-dump-history` if needed. I guess we could move it to `history.bak` though